### ISSUE TITLE
Have a single opam package

### ISSUE THREAD

**Issue by rlepigre** [Deducteam/lambdapi#120]
@ejgallego: do you mind if I remove the lp-lsp package? It don't feel that it makes much sense to have a separate package since we can install the binary together with `lambdapi`. I don't think that such a small tool deserves more than one package on the central Opam repository anyway. What do you think?

If you agree, what should I do with `lp-lsp/README.md`? The version numbers won't make sense with respect to `lambdapi`, as it was not really versioned before...

**Comment by ejgallego** [comment]
I am fine, I will prepare a pull request. Main reason for having a separate package was to allow install of lambdapi with less dependencies, so if the extra deps are ok we can move ahead.

I am always a bit wary of language servers as they tend to depend on lots of stuff. (Lwt / async for example)

On related news I accumulated a larguish Delta in my tree, however is master working for proofs? It seems to me that require is broken so I cannot test.

**Comment by rlepigre** [comment]
OK cool. Master is supposed to work fine, what kind of problem are you having?

**Comment by ejgallego** [comment]
> OK cool. Master is supposed to work fine, what kind of problem are you having?

bool.lp works fine, however in `nat.lp`, I get an error on the `require`.

The error is an `assert false` in `pure.ml`, if I tweak the corresponding line:
```diff
@@ -23,7 +23,9 @@ let parse_text : string -> string -> Command.t Pos.loc list = fun fname s ->
     else Parser.parse_string fname s
   with
   | Fatal(Some(Some(pos)), msg) -> raise (Parse_error(pos, msg))
-  | Fatal(_              , _  ) -> assert false (* Should not produce. *)
+  | Fatal(_              , msg) ->
+    let dummy_pos = Lazy.from_val Pos.{ fname = None; start_line = 0; start_col = 0; end_line = 0; end_col = 0; } in
+    raise (Parse_error(dummy_pos, msg))
```
then I get: `Module [proofs.bool] not loaded (used for binops).` No way to make it work, I tried to compile the stuff, etc...

**Comment by rlepigre** [comment]
Oh, I'll investigate. This may be related to changes I did recently (involving infix operators). I'll investigate.

**Comment by ejgallego** [comment]
Thanks, I was planning to talk about this in person as I felt it would be more effective but now that you mentioned it here we go.

**Comment by rlepigre** [comment]
I know what the problem is! The last line of the `Compile` module initialises a reference in the parser. Indeed, we now need to compile dependencies at parse time in order to fetch the infix symbol that they define. I'm guessing that adding an `open Compile` at the beginning of `src/pure.ml` will solve the problem.

**Comment by ejgallego** [comment]
It would be OK for the LSP server to have to know about files, dependencies among them, and support compilation; the protocol provides facilities for this case.

Maybe it would be better not to compile but to require they are compile. In this case we would need to have a `lpdep` so the LSP server is informed.

**Comment by ejgallego** [comment]
`Compile` does only export `parse_file` tho so a bit of tweaking may be needed, but I think I can advance as I've changed the Pure API a bit on my branch.

**Comment by rlepigre** [comment]
96c74766b8c9da157c434a10fe9ef9a9269ca2a5 fixes part of the problem.

**Comment by rlepigre** [comment]
And indeed, we will need to discuss handling of dependencies. The thing is, even `lpdep` is non-trivial to implement due to infix operators. You don't know how to parse a file until you have loaded its dependencies.

**Comment by ejgallego** [comment]
Based on past and current experiences I would suggest you don't allow `require` outside of the header of a document so `lpdep` can be a fairly simple static parser.

**Comment by rlepigre** [comment]
OK, fair enough. I suggest we discuss that during the Dedukti meeting next week.

**Comment by ejgallego** [comment]
Still getting `Uncaught exception ["Assert_failure src/handle.ml:34:32"]` but indeed we need to see.

**Comment by rlepigre** [comment]
Yeah, as I told you my fix only takes care of part of the problem. I could solve it by attempting to load the file (calling compile) instead of failing, but that is not completely satisfactory.

**Comment by ejgallego** [comment]
Almost there, pity I get now `File [proofs/bool.lpo] is incompatible with current binary`

**Comment by rlepigre** [comment]
Using 8060eee5a306ff4dd13291876e459c9a23c1e65f you should not have assert failures any more. This is not satisfactory, but it suits me as a temporary fix.

For the `.lpo` problem I don't understand. Is this due to the fact that you have old `.lpo` files that were generated with previous version? This error is supposed to only happen if you try to load a `.lpo` with another binary than the one that produced it.

**Comment by ejgallego** [comment]
Umm, that likely means that I cannot use `lambdapi` to compile the files, and `lp-lsp` to load them.

**Comment by rlepigre** [comment]
Yes probably... One thing we could do is build `lp-lsp` into `lambdapi` and have only one binary.

**Comment by ejgallego** [comment]
Well, LP uses `Marshall` to write `Sign.t` where for example `Timed.ref` explicitly states that this is unsafe.

**Comment by ejgallego** [comment]
So that's impossible to debug, but for sure not being able to load object files from different binaries is going to be very cumbersome.

**Comment by rlepigre** [comment]
We actually fall in a case where it is safe to use `Marshall` here, but `Timed.unsafe_reset` must be called on the references, which I apparently don't do (I thought I did).

**Comment by rlepigre** [comment]
But actually, why do you use object files? With my latest change, loading should work.

**Comment by ejgallego** [comment]
I dunno, I was trying to make the mode work so I could test my branch; things work fine now indeed, thanks!

**Comment by rlepigre** [comment]
Note that I fixed the problem with unmarshling with 440b54411222f574399f03d85f885052c5fd24bc, and also a related bug that lead to completely weird behavior in presence of notations.
